game/songparser-txt.cc

```diff
@@ -11,222 +11,222 @@ using namespace SongParserUtil;
-                        double beatDur = getBPM(finalDuet.back().begin).step;
-                        double newEnd = (currentNote.begin - 2*beatDur);
-                        lineBreak.begin = lineBreak.end = newEnd;
-                        if (finalDuet.back().type != Note::SLEEP) {
-                            finalDuet.back().end = newEnd;
-                            if (currentNote.type == Note::SLEEP) { skip = true; }
-                            if (!skip) { finalDuet.push_back(lineBreak); }
-                        }
-                    }
-                }	
-            }
-            if (!skip) { finalDuet.push_back(currentNote); }
-        }
-        auto finalNote = std::unique(finalDuet.begin(), finalDuet.end(), Note::equal);
-        finalDuet.erase(finalNote, finalDuet.end());
-        duetNotes.swap(finalDuet);
-        duetTrack.noteMin = std::min(m_song.getVocalTrack(TrackName::LEAD_VOCAL).noteMin, m_song.getVocalTrack(SongParserUtil::DUET_P2).noteMin);
-        duetTrack.noteMax = std::max(m_song.getVocalTrack(TrackName::LEAD_VOCAL).noteMax, m_song.getVocalTrack(SongParserUtil::DUET_P2).noteMax);
-    } 
+	std::string line;
+	m_curSinger = P1;
+	m_song.insertVocalTrack(TrackName::LEAD_VOCAL, VocalTrack(TrackName::LEAD_VOCAL));
+	m_song.insertVocalTrack(DUET_P2, VocalTrack(DUET_P2));
+	while (getline(line) && txtParseField(line)) {} /// Parse the header again
+	resetNoteParsingState();
+	while (txtParseNote(line) && getline(line)) {} /// Parse notes
+	/// Workaround for the terminating : 1 0 0 line, written by some converters
+	/// FIXME: Should we do this for all tracks?	
+
+	for (auto const& name: { TrackName::LEAD_VOCAL, DUET_P2 }) {
+		Notes& notes = m_song.getVocalTrack(name).notes;
+		auto it = notes.rbegin();
+		if (!notes.empty() && it->type != Note::SLEEP && it->begin == it->end) notes.pop_back();
+		if (notes.empty()) m_song.eraseVocalTrack(name);
+	}
+
+	if (m_song.hasDuet()) {
+		bool skip;
+		Notes s1, s2, merged, finalDuet;
+		s1 = m_song.getVocalTrack(TrackName::LEAD_VOCAL).notes;
+		s2 = m_song.getVocalTrack(SongParserUtil::DUET_P2).notes;
+		std::merge(s1.begin(), s1.end(), s2.begin(), s2.end(), std::back_inserter(merged), Note::ltBegin);
+		VocalTracks const& tracks = m_song.vocalTracks;
+		std::string duetName = tracks.at(TrackName::LEAD_VOCAL).name + " & " + tracks.at(SongParserUtil::DUET_P2).name;
+		m_song.insertVocalTrack(SongParserUtil::DUET_BOTH, duetName);
+		VocalTrack& duetTrack = m_song.getVocalTrack(SongParserUtil::DUET_BOTH);
+		Notes& duetNotes = duetTrack.notes;
+		duetNotes.clear();
+
+		for (auto currentNote: merged) {
+			skip = false;
+			if (!finalDuet.empty()) { 
+				if (currentNote.type == Note::SLEEP) {
+					auto prevToLast = ++(finalDuet.rbegin());
+					if (prevToLast->type == Note::SLEEP) {
+						std::clog << "songparser/info: Phrase formed by a single syllable is most likely our fault, We'll skip the break." << std::endl;
+						skip = true;
+					}
+				}
+				else {
```

Comment:
maybe we should refactor this part to get rid of some if/else clutter.
basically put the success condition at the bottom, and only use if's for exception handling.